### PR TITLE
OHM-51: converted transactions to coffee

### LINKED ISSUE #347: Convert api/transactions.coffee (large) to coffeescript
See epic here: https://github.com/jembi/openhim-core-js/issues/200

### PR DESCRIPTION
**Pull request by devcritter**
Closes #347 

@rcrichton if you could please take a look?

### CODE REVIEW

File: src/api/transactions.coffee [hunk 1]
```diff
@@ -31,259 +32,221 @@ getProjectionObject = (filterRepresentation) ->
...
+
+  # Get the values to use
+  transactionData = this.request.body
+  tx = new transactions.Transaction transactionData
+
+  try
+    # Try to add the new transaction (Call the function that emits a promise and Koa will wait for the function to complete)
+    yield Q.ninvoke tx, "save"
+    this.status = 'created'
+    logger.info 'User %s created transaction with id %s', this.authenticated.email, tx.id
```
Comment (rcrichton) [inline src/api/transactions.coffee]: You use #{} here for consistency but not really a big deal.

File: src/api/transactions.coffee [hunk 2]
```diff
@@ -31,259 +32,221 @@ getProjectionObject = (filterRepresentation) ->
...
+            break
+
+    # --------------Check if user has permission to view full content----------------- #
+    # get projection object
+    projectionFiltersObject = getProjectionObject filterRepresentation
+
+    result = yield transactions.Transaction.findById(transactionId, projectionFiltersObject).exec()
+
+    # Test if the result if valid
+    if result is null or result.length is 0
```
Comment (rcrichton) [inline src/api/transactions.coffee]: You could probably use `result?.length is 0` as above.

File: src/api/transactions.coffee [hunk 3]
```diff
@@ -31,259 +32,221 @@ getProjectionObject = (filterRepresentation) ->
...
+      this.body = "Could not find transaction with ID: #{transactionId}"
+      this.status = 'not found'
+    # Test if the user is authorised
+    else if not authorisation.inGroup 'admin', this.authenticated
+      channels = yield authorisation.getUserViewableChannels(this.authenticated)
```
Comment (rcrichton) [inline src/api/transactions.coffee]: Could get rid of the braces.

File: src/api/transactions.coffee [hunk 4]
```diff
@@ -31,259 +32,221 @@ getProjectionObject = (filterRepresentation) ->
...
+    return
+
+  transactionId = unescape transactionId
+  updates = this.request.body
+
+  try
+    yield transactions.Transaction.findByIdAndUpdate(transactionId, updates).exec()
+    this.body = "Transaction with ID: #{transactionId} successfully updated"
+    this.status = 'ok'
+    logger.info 'User %s updated transaction with id %s', this.authenticated.email, transactionId
```
Comment (rcrichton) [inline src/api/transactions.coffee]: Could use #{} for consitency

File: src/api/transactions.coffee [hunk 5]
```diff
@@ -31,259 +32,221 @@ getProjectionObject = (filterRepresentation) ->
...
+    return
+
+  # Get the values to use
+  transactionId = unescape transactionId
+
+  try
+    yield transactions.Transaction.findByIdAndRemove(transactionId).exec()
+    this.body = 'Transaction successfully deleted'
+    this.status = 'ok'
+    logger.info 'User %s removed transaction with id %s', this.authenticated.email, transactionId
```
Comment (rcrichton) [inline src/api/transactions.coffee]: Could use #{} for consitency

**Comment by rcrichton** [comment]
Hey @devcritter looks good :+1: Just a few comments above, all minor. The test seems to be failing but not from you you have done. I'm re-running them now.

File: src/api/transactions.coffee (hunk 2 above)
Comment (devcritter) [inline src/api/transactions.coffee]: Cool, yeah I missed this one

**Pull request merged**